### PR TITLE
dtoh: Properly emit class templates

### PR DESCRIPTION
**Pull request by MoonlightSentinel**
Properly emit class template declarations without segfaulting& creating invalid declarations.

---

~~Needs some more work because the AST is really inconsistent for templated `ClassDeclaration`s~~

### CODE REVIEW

File: src/dmd/frontend.h [hunk 1]
```diff
@@ -6251,6 +6258,245 @@ enum class PREC
     primary = 16,
 };
 
+template <typename AST>
+class ParseTimeVisitor
+{
+public:
+    void visit(typename AST::Dsymbol );
```
Comment (ibuclaw) [inline src/dmd/frontend.h:6265]: Seems to be missing virtual.

**Comment by WalterBright** [comment]
Just a comment on the subject: "dtoh: properly handle templates" conveys little to no information, might as well be "something about dtoh and templates". Can we improve this?

**Comment by MoonlightSentinel** [comment]
This is a **Ddraft PR** to check the implementation on on different targets. I'll extract actuall pull requests as stated in the description (see e.g. #11877, ...)

File: src/dmd/frontend.h (hunk 1 above)
Comment (MoonlightSentinel) [reply src/dmd/frontend.h:6265]: Seems as if that information is missing in the generated AST....

Comment (ibuclaw) [reply src/dmd/frontend.h:6265]: No `vtblIndex`?

File: src/dmd/dtoh.d [hunk 2]
```diff
@@ -1778,39 +1802,44 @@ public:
...
-        {
-            buf.writestring(sd.isUnionDeclaration() ? "union " : "struct ");
-            buf.writestring(sd.ident.toChars());
-            if (sd.members)
+            if (p.isTemplateTypeParameter())
             {
-                buf.writenl();
-                buf.writestringln("{");
-                auto savex = adparent;
-                adparent = sd;
```
Comment (ibuclaw) [inline src/dmd/dtoh.d:1795]: @MoonlightSentinel it's because you removed this line that the functions aren't written out as virtual.

Comment (ibuclaw) [reply src/dmd/dtoh.d:1795]: Though looking at the old code, it only set this for templated structs/unions, not classes.

Besides, the condition for whether or not to emit `virtual` shouldn't depend on whether `adparent` is set.  `fd.isVirtual() && fd.vtblIndex != -1` is how other parts of the compiler deal with it.  (`adparent` is used for the checkbuf assertion however).

Comment (MoonlightSentinel) [reply src/dmd/dtoh.d:1795]: I see, thanks!
Not sure why `dtoh` even checks `adparent` instead of using `fd.isVirtual()`/ `fd.isVirtualFunction()`

**Comment by MoonlightSentinel** [comment]
Dedicated this PR to `ClassDeclaration`s because most of the other changes were merged already.

Needs some more work w.r.t `virtual`.

File: src/dmd/dtoh.d (hunk 2 above)
Comment (MoonlightSentinel) [reply src/dmd/dtoh.d]: > Besides, the condition for whether or not to emit virtual shouldn't depend on whether adparent is set.

Ideally not, but `vtblIndex` is still `-1` for templated `ClassDeclaration`s.

File: src/dmd/dtoh.d [hunk 3]
```diff
@@ -257,6 +257,7 @@ public:
     bool forwardedAA;
     AST.Type* origType;
     AST.Prot.Kind currentProt; /// Last written protection level
+    ulong storageClass; /// Currently applicable storage classes
```
Comment (thewilsonator) [inline src/dmd/dtoh.d:260]: `STC storageClass`?

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/dtoh.d (hunk 3 above)
Comment (MoonlightSentinel) [reply src/dmd/dtoh.d]: Fixed

File: src/dmd/frontend.h [hunk 4]
```diff
@@ -5814,6 +5816,245 @@ enum class PREC
     primary = 16,
 };
 
+template <typename AST>
+class ParseTimeVisitor
+{
+public:
+    virtual void visit(typename AST::Dsymbol );
```
Comment (ibuclaw) [inline src/dmd/frontend.h:5823]: Are you sure this is right?

**Comment by WalterBright** [comment]
@MoonlightSentinel the subject was never fixed as I requested.

**Pull request merged**